### ISSUE TITLE
Instanced Scripts

### ISSUE THREAD

**Issue by Tanjitsu** [crosire/scripthookvdotnet#659]
Stored in this [Branch](https://github.com/Tanjitsu/scripthookvdotnet/tree/dev_v3_instance)  Relating to [issues/545](https://github.com/crosire/scripthookvdotnet/issues/545)   Started in [pull/657](https://github.com/crosire/scripthookvdotnet/pull/657)

New [Branch](https://github.com/Tanjitsu/scripthookvdotnet/tree/AddScript) add for Rebuilt ASI in Master Branch.
*Branch updated with "Example" script.

**Comment by Tanjitsu** [comment]
I have updated the branch with a working ScriptInstance.Wait() system.

new ScriptInstance();
ScriptInstance.IsRegistered() ?
Script.Register(ScriptInstance);

ScriptInstance.Start()
ScriptInstance.Started() ?
ScriptInstance.Pause()
EventHandler Paused;

ScriptInstance.Abort();

**Comment by Tanjitsu** [comment]
@crosire If you get chance , compile / look over the branch.
I have updated the first comments Example with a basic starting point.

**Comment by Tanjitsu** [comment]
[03:06:36] [INFO] Found 1 script(s) with 12 instance(s) in....

My current setup has been moved over to use the instance branch as per above.
Seems to work as expected...

**Comment by Tanjitsu** [comment]
If anyone else tries this... feed back would be good.

**Comment by crosire** [comment]
Sorry, got a lot going on right now. I'll try to check this out during the weekend.

**Comment by Tanjitsu** [comment]
@crosire Coolio, dont mean to be a pest just want to see things moving forward :)

**Comment by crosire** [comment]
I'd prefer if instances can be created from normal `Script` classes instead of introducing another `ScriptInstance` type (to avoid additional complexity). We could add a `NoDefaultInstance` property to the `ScriptAttributes` class attribute so you can prevent SHVDN from creating a default instance of a script (like it does right now) and instead handle creation/aborting yourself from some main script. This has the added benefit that it would technically allow the main script (any script really) to abort itself too, which could be useful.

Example:
```c#
class MainScript : Script
{
	WorkerScript instance;

	public MainScript()
	{
		KeyDown +=  OnKeyDown;
		instance = Script.CreateInstance<WorkerScript>();
	}

	private void OnKeyDown(object sender, KeyEventArgs e)
	{
		if (e.KeyCode == Keys.A)
			instance.Start();
		if (e.KeyCode == Keys.B)
			instance.Abort();
	}
}

[ScriptAttributes(NoDefaultInstance = true)]
class WorkerScript : Script
{
}
```

**Comment by Tanjitsu** [comment]
Ok.  Is "NoDefaultInstance" readable from within the instance ?

Script.Wait() is static, 
Instance.Wait() is not.

You can call Script.Wait from anywhere and the Executing thread will wait.
If you call Instance.Wait. it will queue the wait until the Instance is Executing.
(unless the Instance is the currently Executing thread, then it will wait immediately).

**Comment by crosire** [comment]
It's a class attribute, you can get that with reflection. Though of you add it to a class you already know that it's there, so I don't see why you would have to read it?! 

What is the usecase for forcing another script to wait? That sounds really dangerous since the code issuing a wait command for another instance does not know which codepath that instance is currently executing.

**Comment by Tanjitsu** [comment]
> though of you add it to a class you already know that it's there,

The Domain handling the "Script" would need to know it was there so it could handle it as an Instance or a Script. 

> What is the usecase for forcing another script to wait? 

Try the example... 
Script 1. Tick()
create instance....
Call instance.Wait....  <- not in "Instance" so sets wait.
End Tick();

Instance.Tick()
Wait isSet Do.Wait <-  is in "instance" so begins wait.
End Tick();

Instance.Tick();
Call Wait <-  is in "instance" so begins wait.
End Tick();

Im not a Threading expert :)

**Comment by RainerHilmer** [comment]
Please correct me if I'm wrong but this is about interaction between two scripts (or one main script and another sub-script), right? Here is what I do: For example there is my Teleportals mod which involes severals scripts working parallel and with different tick intervals. One script (Preselection) fires just every 500ms for querying nearby portals. Another script is solely for rendering markers each frame. Both communicate through (what I call) a Mediator (don't confuse with the Mediator pattern). This is a simple static class with properties like bool SomethingSomethingActive (take whatever name you like). One script sets it to true and as long as this property is in this state, the other script waits (better say the event handler just returns w/o any action besides querying this property). Works like a charm.

**Comment by Tanjitsu** [comment]
@RainerHilmer  No. This is creating new scripts on the fly from a static script.

> *Branch updated with "Example" script.

You will still use a static mediator when talking between threads but the point is to create threads as and when you need. Not predefined on start up.

I have 1 Script "Manager"
Which then as it needs creates as many or as few of my other 14 InstanceScripts, when ever my system requires them.

So i have 1 InstanceScript called "member", the system may have upto 7 Instances of "member" running as per my manager.

**Comment by RainerHilmer** [comment]
Wow that's great!

**Comment by Tanjitsu** [comment]
I have been using now for a month and seem to have no issues once my system was re-tailored to work dynamically. I would like to re-work the system to use bool flags better for pausing / aborting the instances cleaner. Based on crosire's comments when i do ill try to do it just using "Script" and not using the second class definition of "ScriptInstance".

**Comment by RainerHilmer** [comment]
So my comment was a bit helpful?

**Comment by Tanjitsu** [comment]
The static mediator comment wasn't really needed as that's a fairly basic principle to follow when working with none static values but it's nice to see someone is interested in the thread.

I thought alot of people where wanting to be able to Instance threads on demand instead of all predefined.
It allows simple thing like this

var notification = UI.Notify("blah");
Wait(2000);
notification.Hide();

:) Or far more complex scripts to be called and disposed of, as per your needs.

**Comment by DurtyFree** [comment]
Why is this closed? This looks like a more convenient way of "threading" scripts

**Comment by kagikn** [comment]
Yeah, this must be worth adding. RAGE Plugin Hook has GameFiber class, which enables us to run multiple thread in one dll. http://docs.ragepluginhook.net/html/T_Rage_GameFiber.htm
![screenshot_2017-10-11-17-40-57](https://user-images.githubusercontent.com/9353978/31430302-73caace0-aeab-11e7-9dc5-141ce236d563.png)

**Comment by DurtyFree** [comment]
Maybe @Tanjitsu is able to create a new pull request, would be really worth adding.

This could really increase the performance of some mods if its properly used. #620 
Would love to get this added.

**Comment by crosire** [comment]
This would not increase performance. It would only allow multiple instances of the same script, which still run through the same pipeline.

**Comment by DurtyFree** [comment]
@crosire 
We have split natives run on our mod per tick to different scripts and we experienced an increase of the FPS.

**Comment by crosire** [comment]
SHVDN still serializes those. Scripts/Instances do **not** run in patallel in SHVDN.

**Comment by DurtyFree** [comment]
Good to know, so do you think the approach of seperating a mod into multiple scripts in order to gain FPS is a waste of time or even problematic in some cases?

**Comment by crosire** [comment]
Not necessarily. If you do a lot of CPU work between calling natives, then yes, multiple scripts help. If you only call lots of natives, then no, it won't. I should have clarified: The native function execution is serialized, but C# code does run in parallel.

**Comment by RainerHilmer** [comment]
Just use a [Balking pattern](https://en.wikipedia.org/wiki/Balking_pattern) for the communication between multiple scripts. That way a fast ticking handler doesn't get halted by slower scripts. I use this approach in my Teleportals mod and it works flawless.

**Comment by Tanjitsu** [comment]
> This would not increase performance. It would only allow multiple instances of the same script, which still run through the same pipeline.

That's not true, managing dynamic threads instead of static threads can provide alot of performance increases. It is dependant upon your method of handling your threads / script content.

The current "static" method is restrictive and so basic.
The ability to call and abort threads on demand unlocks "moders / coders" to be able to code freely.
but then i forget that DotNet is aimed at the public not at intelligent people.

**Comment by Tanjitsu** [comment]
I got chance to update the Instance branch to match the latest changes.
[Script Instance](https://github.com/Tanjitsu/scripthookvdotnet/tree/dev_v3_instance)

I simply updated it so it still works on the old method of a second class "ScriptInstance".

**Comment by Tanjitsu** [comment]
I have reworked this for the new Master branch under [This Branch](https://github.com/Tanjitsu/scripthookvdotnet/tree/AddScript).

This is now an extension instead of an alteration.
Using Script Attribute for Script AutoStart.
Example Script added for testing.

**Comment by DurtyFree** [comment]
Can this please be an official feature of SHVDN?

**Comment by crosire** [comment]
This has been merged now (#886).
